pgjdbc/src/main/java/org/postgresql/jdbc/Arrays.java

```diff
@@ -0,0 +1,1412 @@
+      Oid.BOOL, Oid.BOOL_ARRAY) {
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void appendArray(StringBuilder sb, char delim, boolean[] array) {
+      sb.append('{');
+      for (int i = 0; i < array.length; ++i) {
+        if (i > 0) {
+          sb.append(delim);
+        }
+        sb.append(array[i] ? '1' : '0');
+      }
+      sb.append('}');
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    protected void write(boolean[] array, byte[] bytes, int offset) {
+      int idx = offset;
+      for (int i = 0; i < array.length; ++i) {
+        bytes[idx + 3] = 1;
+        ByteConverter.bool(bytes, idx + 4, array[i]);
+        idx += 5;
+      }
+    }
+  };
+
+  private static final AbstractArraySupport<Boolean[]> BOOLEAN_OBJ_ARRAY = new AbstractArraySupport<Boolean[]>(Oid.BOOL,
+      Oid.BOOL_ARRAY) {
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public byte[] toBinaryRepresentation(BaseConnection connection, Boolean[] array, int oid)
+        throws SQLException, SQLFeatureNotSupportedException {
+      assert oid == arrayOid;
+
+      final int nullCount = countNulls(array);
+
+      final byte[] bytes = writeBytes(array, nullCount, 20);
+
+      // 1 dimension
+      ByteConverter.int4(bytes, 0, 1);
+      // no null
+      ByteConverter.int4(bytes, 4, nullCount == 0 ? 0 : 1);
+      // oid
+      ByteConverter.int4(bytes, 8, getTypeOID(oid));
+      // length
+      ByteConverter.int4(bytes, 12, array.length);
+
+      return bytes;
+    }
+
+    private byte[] writeBytes(final Boolean[] array, final int nullCount, final int offset) {
+      final int length = offset + (4 * array.length) + (array.length - nullCount);
+      final byte[] bytes = new byte[length];
+
+      int idx = offset;
+      for (int i = 0; i < array.length; ++i) {
+        if (array[i] == null) {
+          ByteConverter.int4(bytes, idx, -1);
+          idx += 4;
+        } else {
+          ByteConverter.int4(bytes, idx, 1);
+          idx += 4;
+          write(array[i], bytes, idx);
+          ++idx;
+        }
+      }
+
+      return bytes;
+    }
+
+    private void write(Boolean bool, byte[] bytes, int idx) {
+      ByteConverter.bool(bytes, idx, bool.booleanValue());
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    byte[] toSingleDimensionBinaryRepresentation(BaseConnection connection, Boolean[] array)
+        throws SQLException, SQLFeatureNotSupportedException {
+      final int nullCount = countNulls(array);
+      return writeBytes(array, nullCount, 0);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    void appendArray(StringBuilder sb, char delim, Boolean[] array) {
+      sb.append('{');
+      for (int i = 0; i < array.length; ++i) {
+        if (i != 0) {
+          sb.append(delim);
+        }
+        if (array[i] == null) {
+          sb.append('N');
```

Comment:
This seems intentional, but awkward. What is the benefit of appending each character separately instead of appending the whole string?

Reply:
how would you put nulls in then ?

Reply:
Got it, makes sense to avoid the string internals. Since we're micro optimizing, you could chain the ‘.append('N').append('U')...’ calls. It results in smaller byte code which can make the difference between a method being inlined or not.

Reply:
done